"Travis workarounds" should be enabled for all Windows targets when GNU toolchains are used

Not sure what happened but my computer forced the creation of this ticket before I was done. So this looks weird.

Issue #229 Discusses this same problem where llvm-config does not exist. This is because llvm-config is NOT included by default in the llvm bprebuilt binaries for windows. The reason for this is they don't want you using llvm as a library on windows, only a toolchain.

Therefore the instructions provide won't work for windows users. They either need to build llvm from scratch, or the dependency on llvm-config needs to be removed (preferably the second in my opinion.)

Heh. I was missing some other dependencies, so I think it was crapping out initially making it seem like it was just llvm-config. After I fixed that I get this:

```
PS F:\Godot\rust-test> cargo build
   Compiling gdnative-sys v0.8.1
The following warnings were emitted during compilation:

warning: couldn't execute `llvm-config --prefix` (error: The system cannot find the file specified. (os error 2))
warning: set the LLVM_CONFIG_PATH environment variable to a valid `llvm-config` executable

error: failed to run custom build command for `gdnative-sys v0.8.1`

Caused by:
  process didn't exit successfully: `F:\Godot\rust-test\target\debug\build\gdnative-sys-b393d5fc5aaba90c\build-script-build` (exit code: 101)
--- stdout
cargo:warning=couldn't execute `llvm-config --prefix` (error: The system cannot find the file specified. (os error 2))
cargo:warning=set the LLVM_CONFIG_PATH environment variable to a valid `llvm-config` executable

--- stderr
C:\Program Files (x86)\Microsoft Visual Studio\2019\Community\VC\Tools\MSVC\14.26.28801\include\vadefs.h:61:33: error: expected ';' after top level declarator
C:\Program Files (x86)\Microsoft Visual Studio\2019\Community\VC\Tools\MSVC\14.26.28801\include\vcruntime.h:96:1: error: expected identifier
C:\Program Files (x86)\Microsoft Visual Studio\2019\Community\VC\Tools\MSVC\14.26.28801\include\vcruntime.h:96:20: error: expected ';' after top level declarator
C:\Program Files (x86)\Microsoft Visual Studio\2019\Community\VC\Tools\MSVC\14.26.28801\include\vcruntime.h:193:29: error: expected ';' after top level declarator
C:\Program Files (x86)\Microsoft Visual Studio\2019\Community\VC\Tools\MSVC\14.26.28801\include\vcruntime.h:344:51: error: unknown type name 'uintptr_t'
C:\Program Files (x86)\Microsoft Visual Studio\2019\Community\VC\Tools\MSVC\14.26.28801\include\vcruntime.h:345:67: error: unknown type name 'uintptr_t'
C:\Program Files (x86)\Microsoft Visual Studio\2019\Community\VC\Tools\MSVC\14.26.28801\include\vcruntime.h:349:8: error: unknown type name 'uintptr_t'
C:\Program Files (x86)\Microsoft Visual Studio\2019\Community\VC\Tools\MSVC\14.26.28801\include\vcruntime.h:357:1: error: a parameter list without types is only allowed in a function definition
C:\Program Files (x86)\Microsoft Visual Studio\2019\Community\VC\Tools\MSVC\14.26.28801\include\vcruntime.h:357:18: error: expected ';' after top level declarator
C:\Program Files (x86)\Windows Kits\10\Include\10.0.18362.0\ucrt\corecrt.h:12:1: error: expected identifier
C:\Program Files (x86)\Windows Kits\10\Include\10.0.18362.0\ucrt\corecrt.h:12:20: error: expected ';' after top level declarator
C:\Program Files (x86)\Windows Kits\10\Include\10.0.18362.0\ucrt\corecrt.h:287:16: error: unknown type name 'uintptr_t'
C:\Program Files (x86)\Windows Kits\10\Include\10.0.18362.0\ucrt\corecrt.h:562:13: error: unknown type name 'size_t'
C:\Program Files (x86)\Windows Kits\10\Include\10.0.18362.0\ucrt\corecrt.h:2011:1: error: a parameter list without types is only allowed in a function definition
C:\Program Files (x86)\Windows Kits\10\Include\10.0.18362.0\ucrt\errno.h:14:1: error: expected function body after function declarator
C:\Program Files (x86)\Windows Kits\10\Include\10.0.18362.0\ucrt\errno.h:130:1: error: a parameter list without types is only allowed in a function definition
C:\Program Files (x86)\Microsoft Visual Studio\2019\Community\VC\Tools\MSVC\14.26.28801\include\vcruntime_string.h:17:1: error: expected function body after function declarator
C:\Program Files (x86)\Microsoft Visual Studio\2019\Community\VC\Tools\MSVC\14.26.28801\include\vcruntime_string.h:32:29: error: unknown type name 'size_t'
C:\Program Files (x86)\Microsoft Visual Studio\2019\Community\VC\Tools\MSVC\14.26.28801\include\vcruntime_string.h:46:35: error: unknown type name 'size_t'
fatal error: too many errors emitted, stopping now [-ferror-limit=]
C:\Program Files (x86)\Microsoft Visual Studio\2019\Community\VC\Tools\MSVC\14.26.28801\include\vadefs.h:61:33: error: expected ';' after top level declarator, err: true
C:\Program Files (x86)\Microsoft Visual Studio\2019\Community\VC\Tools\MSVC\14.26.28801\include\vcruntime.h:96:1: error: expected identifier, err: true
C:\Program Files (x86)\Microsoft Visual Studio\2019\Community\VC\Tools\MSVC\14.26.28801\include\vcruntime.h:96:20: error: expected ';' after top level declarator, err: true
C:\Program Files (x86)\Microsoft Visual Studio\2019\Community\VC\Tools\MSVC\14.26.28801\include\vcruntime.h:193:29: error: expected ';' after top level declarator, err: true
C:\Program Files (x86)\Microsoft Visual Studio\2019\Community\VC\Tools\MSVC\14.26.28801\include\vcruntime.h:344:51: error: unknown type name 'uintptr_t', err: true
C:\Program Files (x86)\Microsoft Visual Studio\2019\Community\VC\Tools\MSVC\14.26.28801\include\vcruntime.h:345:67: error: unknown type name 'uintptr_t', err: true
C:\Program Files (x86)\Microsoft Visual Studio\2019\Community\VC\Tools\MSVC\14.26.28801\include\vcruntime.h:349:8: error: unknown type name 'uintptr_t', err: true
C:\Program Files (x86)\Microsoft Visual Studio\2019\Community\VC\Tools\MSVC\14.26.28801\include\vcruntime.h:357:1: error: a parameter list without types is only allowed in a function definition, err: true
C:\Program Files (x86)\Microsoft Visual Studio\2019\Community\VC\Tools\MSVC\14.26.28801\include\vcruntime.h:357:18: error: expected ';' after top level declarator, err: true
C:\Program Files (x86)\Windows Kits\10\Include\10.0.18362.0\ucrt\corecrt.h:12:1: error: expected identifier, err: true
C:\Program Files (x86)\Windows Kits\10\Include\10.0.18362.0\ucrt\corecrt.h:12:20: error: expected ';' after top level declarator, err: true
C:\Program Files (x86)\Windows Kits\10\Include\10.0.18362.0\ucrt\corecrt.h:287:16: error: unknown type name 'uintptr_t', err: true
C:\Program Files (x86)\Windows Kits\10\Include\10.0.18362.0\ucrt\corecrt.h:562:13: error: unknown type name 'size_t', err: true
C:\Program Files (x86)\Windows Kits\10\Include\10.0.18362.0\ucrt\corecrt.h:2011:1: error: a parameter list without types is only allowed in a function definition, err: true
C:\Program Files (x86)\Windows Kits\10\Include\10.0.18362.0\ucrt\errno.h:14:1: error: expected function body after function declarator, err: true
C:\Program Files (x86)\Windows Kits\10\Include\10.0.18362.0\ucrt\errno.h:130:1: error: a parameter list without types is only allowed in a function definition, err: true
C:\Program Files (x86)\Microsoft Visual Studio\2019\Community\VC\Tools\MSVC\14.26.28801\include\vcruntime_string.h:17:1: error: expected function body after function declarator, err: true
C:\Program Files (x86)\Microsoft Visual Studio\2019\Community\VC\Tools\MSVC\14.26.28801\include\vcruntime_string.h:32:29: error: unknown type name 'size_t', err: true
C:\Program Files (x86)\Microsoft Visual Studio\2019\Community\VC\Tools\MSVC\14.26.28801\include\vcruntime_string.h:46:35: error: unknown type name 'size_t', err: true
fatal error: too many errors emitted, stopping now [-ferror-limit=], err: true
thread 'main' panicked at 'Unable to generate bindings: ()', C:\Users\seanm\scoop\persist\rustup\.cargo\registry\src\github.com-1ecc6299db9ec823\gdnative-sys-0.8.1\build.rs:217:24
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
```

--- a/gdnative-sys/build.rs
+++ b/gdnative-sys/build.rs
@@ -153,16 +153,6 @@ mod header_binding {
         builder
     }
 
-    fn is_travis_ci() -> bool {
-        if let Ok(is_ci) = std::env::var("CI") {
-            if let Ok(is_travis) = std::env::var("TRAVIS") {
-                return is_ci == "true" && is_travis == "true";
-            }
-        }
-
-        false
-    }
-
     #[allow(clippy::single_match)]
     pub(crate) fn generate(manifest_dir: &str, out_dir: &str) {
         // on mac/iOS this will be modified, so it is marked as mutable.
@@ -183,6 +173,7 @@ mod header_binding {
         let target_os = std::env::var("CARGO_CFG_TARGET_OS").unwrap();
         let target_arch = std::env::var("CARGO_CFG_TARGET_ARCH").unwrap();
         let target_vendor = std::env::var("CARGO_CFG_TARGET_VENDOR").unwrap();
+        let target_env = std::env::var("CARGO_CFG_TARGET_ENV").unwrap();
 
         if target_vendor == "apple" {
             match apple_include_path() {
@@ -199,21 +190,25 @@ mod header_binding {
             builder = builder.clang_arg("--target=arm64-apple-ios");
         }
 
-        // Workaround: Somehow, Microsoft extensions aren't enabled by default on Travis's
-        // Windows environment. We need to enable it manually, or MSVC headers will fail to
-        // parse. We also need to manually define architecture macros, or the build will fail
-        // with an "Unsupported architecture" error.
+        // Workaround: Microsoft extensions aren't enabled by default for the `gnu` toolchain
+        // on Windows. We need to enable it manually, or MSVC headers will fail to parse. We
+        // also need to manually define architecture macros, or the build will fail with an
+        // "Unsupported architecture" error.
         //
-        // This does not happen with "normal" Windows environments.
-        if target_os == "windows" && is_travis_ci() {
-            if target_arch != "x86_64" {
-                panic!("Windows environment on Travis CI should be x86_64")
-            }
+        // This does not happen when the `msvc` toolchain is used.
+        if target_os == "windows" && target_env == "gnu" {
+            let arch_macro = match target_arch.as_str() {
+                "x86" => "_M_IX86",
+                "x86_64" => "_M_X64",
+                "arm" => "_M_ARM",
+                "aarch64" => "_M_ARM64",
+                _ => panic!("architecture {} not supported on Windows", target_arch),
+            };
 
             builder = builder
                 .clang_arg("-fms-extensions")
                 .clang_arg("-fmsc-version=1300")
-                .clang_arg("-D_M_X64=100");
+                .clang_arg(format!("-D{}=100", arch_macro));
         }
 
         if target_os == "android" {

